### DISCUSSION TITLE
FAPI-SIG tool for tests 'FAPI1 - Advanced-Final: Brazil Dynamic Client Registration Authorization server test'

### DISCUSSION THREAD

**Discussion by fassumpcao** [keycloak/keycloak#10422] [Q&A]
We are trying to used FAPI-SIG tool for tests 'FAPI1 - Advanced-Final: Brazil Dynamic Client Registration Authorization server test' with MTLS Authentication.
We add success a new client using gateway for request a Keycloak registration endpoint. But, in the next step, that we need to receive an access_token with transport certificate (Certificate generate in Open Banking Brazil Directory), we receive a handshake error and the request stopped in load_balancer service. 
When we tryied to request using 'client' or 'client2' certificates, the request make success on Keycloak.
I would like to someone help us to apply a correct configuration of server certificates on FAPI-SIG tools using Open Banking certificate.

**Comment by tnorimat** [comment]
@fassumpcao Hello,

> we receive a handshake error and the request stopped in load_balancer service.

Could you tell me which kind of an error happens in detail?

**Reply by tnorimat** [reply]
@fassumpcao By the way, the current FAPI-SIG's automated conformance test run environment does not support Open Banking Brazil FAPI 1.0 Dynamic Client Registration conformance test while it supports other all conformance tests.

**Comment by fassumpcao** [comment]
Hello Takashi,
I know that, but we created a gateway to extract software_statement of the
origin DCR request  and send to Keycloak registration endpoint, and it's
working. But, in the next step, to get an access_token, we receive:

Using Curl:
curl: (35) error:14094416:SSL routines:ssl3_read_bytes:sslv3 alert
certificate unknown

Using Postman:
Error: write EPROTO 51945928:error:10000416:SSL
routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_UNKNOWN:../../third_party/boringssl/src/ssl/tls_record.cc:594:SSL
alert number 46

In other authorization servers from other Financial Institutions, we did
the DCR step and the token using the same certificates and it worked.
____________________________________________________
______________________________
*Felipe P. de Assumpção*


Em qua., 23 de fev. de 2022 às 20:04, Takashi Norimatsu <
***@***.***> escreveu:

**Reply by tnorimat** [reply]
I'm not familiar with OBB DCR specification but I'm afraid that keycloak fails to verify a client certificate during mutual TLS handshare.
I expect that keycloak does not hold an appropriate CA certificate (may be Brazil ICP ?) needed to verify a client certificate (may be issued by Brazil ICP ?) during mutual TLS handshare.

The following codes setup keycloak's truststore:
keycloak/Dockerfile:
```
# Add certificates
COPY server.pem /etc/x509/https/server.crt
COPY client-ca.pem /etc/x509/https/client-ca.crt

RUN \
  echo "Importing certificates into truststore" && \
  keytool -import -cacerts -storepass changeit -noprompt -file /etc/x509/https/server.crt -alias myservercrt && \
  keytool -import -cacerts -storepass changeit -noprompt -file /etc/x509/https/client-ca.crt  -alias myclientcacrt && \
  echo "Certificates imported"
```
How about adding an appropriate CA certificate (may be Brazil ICP ?) to this truststore?

**Reply by tnorimat** [reply]
If you use wildfly based keycloak, how about adding an appripriate CA certificate to the following environment variable?

keycloak/Dockerfile.legacy:
```
ENV X509_CA_BUNDLE="/etc/x509/https/client-ca.crt /etc/x509/https/server.crt"
```
